### PR TITLE
Adding the draft patch for tblgen type definitions

### PR DESCRIPTION
**Pull request by teqdruid**
https://reviews.llvm.org/D86904

### CODE REVIEW

**Comment by teqdruid** [comment]
Requesting a brief review since this patch hasn't made it into LLVM yet

**Review by mikeurbach — APPROVED** [review]

File: mlir/include/mlir/TableGen/TypeDef.h [hunk 1]
```diff
@@ -0,0 +1,142 @@
+//
```
Comment (lattner) [inline mlir/include/mlir/TableGen/TypeDef.h:1]: It looks like this lost the first line of the header comment.

File: mlir/include/mlir/IR/OpBase.td [hunk 2]
```diff
@@ -2351,4 +2351,92 @@ def location;
...
+  let allocator = [{$_dst = $_allocator.copyInto($_self);}];
+}
+
+// For classes which require allocation and have their own allocateInto method
+class SelfAllocationParameter<string type, string desc> : TypeParameter<type, desc> {
+  let allocator = [{$_dst = $_self.allocateInto($_allocator);}];
+}
+
+// For ArrayRefs which contain things which allocate themselves
+class ArrayRefOfSelfAllocationParameter<string arrayOf, string desc> : TypeParameter<"::llvm::ArrayRef<" # arrayOf # ">", desc> {
```
Comment (lattner) [inline mlir/include/mlir/IR/OpBase.td:2431]: Please fit in 80 columns

File: mlir/include/mlir/IR/OpBase.td [hunk 3]
```diff
@@ -2351,4 +2351,92 @@ def location;
...
+  let allocator = [{
+    llvm::SmallVector<}] # arrayOf # [{, 4> tmpFields;
+    for (size_t i = 0; i < $_self.size(); i++) {
```
Comment (lattner) [inline mlir/include/mlir/IR/OpBase.td:2434]: No need to evaluate .size() each time through the loop, prefer something like:

```
for (size_t i = 0, e = $_self.size(); i != e; ++i)
```

File: mlir/include/mlir/IR/OpBase.td [hunk 4]
```diff
@@ -2351,4 +2351,92 @@ def location;
...
+  string syntax = ?;
+}
+
+// For StringRefs, which require allocation
+class StringRefParameter<string desc> : TypeParameter<"::llvm::StringRef", desc> {
+  let allocator = [{$_dst = $_allocator.copyInto($_self);}];
+}
+
+// For standard ArrayRefs, which require allocation
+class ArrayRefParameter<string arrayOf, string desc> : TypeParameter<"::llvm::ArrayRef<" # arrayOf # ">", desc> {
```
Comment (lattner) [inline mlir/include/mlir/IR/OpBase.td:2421]: 80 columns plz

File: mlir/test/lib/Dialect/Test/TestTypes.cpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
+//===- TestTypes.cpp - MLIR Test Dialect Types ----------------*- C++ -*-===//
```
Comment (lattner) [inline mlir/test/lib/Dialect/Test/TestTypes.cpp:1]: Plz fill to 80 columns.

File: mlir/tools/mlir-tblgen/TypeDefGen.cpp [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+#include "mlir/TableGen/Format.h"
+#include "mlir/TableGen/GenInfo.h"
+#include "mlir/TableGen/Interfaces.h"
+#include "mlir/TableGen/OpClass.h"
+#include "mlir/TableGen/OpTrait.h"
+#include "mlir/TableGen/Operator.h"
+#include "mlir/TableGen/TypeDef.h"
+#include "llvm/ADT/Sequence.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/StringExtras.h"
```
Comment (lattner) [inline mlir/tools/mlir-tblgen/TypeDefGen.cpp:24]: Is it possible to trim out any of these #includes?

**Review by lattner — COMMENTED** [review]
I'm a bit nervous about this going into the CIRCT tree before the MLIR tree because it will make future merges more complicated, but I assume you're working on upstreaming this as well :)

**Comment by teqdruid** [comment]
@lattner: [actively](https://reviews.llvm.org/D86904) but it's taking longer to review than I would like and it's gating some PRs into llvm/circt. (Which is natural for a feature of this size.) We've already gone through a few revisions. I've already gotten a preparation patch into MLIR which changes the files which have tended to have the most merge conflicts. We discussed this on discord, which I know you're not on. Sorry.

The thinking is that all the small stuff would be taken care of in the patch which eventually makes its way into MLIR upstream. Once it's landed (and assuming no other CIRCT-specific patches are present), we just ditch this change and start following llvm/master again, by just resetting the master branch to match theirs. (This seems to be what was done for Stephen's RegionKind patch AFAICT.)

Given that this PR has already been merged, I don't think it's worth revisiting (since none of them are functional changes) but I will update my patch into MLIR based on these comments! This is "short"-term anyway.

**Comment by teqdruid** [comment]
@lattner Why am I being lazy? Easy enough to apply a cleanup patch in both places. https://github.com/circt/llvm/pull/4

**Comment by lattner** [comment]
Cool, thanks @teqdruid

**Pull request merged**